powerline/bindings/bar/powerline-bar.py

```diff
@@ -35,18 +31,37 @@ def render(event=None, data=None, sub=None):
 		help='Subscribe for i3 events.'
 	)
 	args = parser.parse_args()
-	powerline = BarPowerline()
-
-	interval = 0.5
+	line = BarPowerline()
 	lock = Lock()
-
 	write = get_unicode_writer(encoding='utf-8')
 
+	def render(reschedule=False):
+		if reschedule:
+			Timer(0.5, render, kwargs={"reschedule": True}).start()
+
+		global lock
+		with lock:
+			write(line.render())
+			write('\n')
+			sys.stdout.flush()
+
+	def update(mode):
+		line.mode = mode
+		render()
+
+	render(reschedule=True)
+
 	if args.i3:
-		import i3
-		sub = i3.Subscription(render, 'workspace')
+		try:
+			import i3ipc
+		except ImportError:
+			import i3
+			i3.Subscription(lambda evt, data, sub: print(render()), 'workspace')
+		else:
+			conn = i3ipc.Connection()
+			conn.on('workspace::focus', lambda conn, evt: render())
+			conn.on('mode', lambda conn, evt: update(evt.change))
+			conn.main()
 
 	while True:
-		start_time = monotonic()
-		render()
-		time.sleep(max(interval - (monotonic() - start_time), 0.1))
+		time.sleep(0.5)
```

Comment:
Why are you using `0.5` here? You don’t depend on the interval here, so you can use any great number like `1e10`, job will be done in other threads in any case and sleeping does not affect signal processing.

This value looks like if you *need* 0.5 for some reason because you are using it in the `Timer` call which calls for a refactoring “replace two magic numbers with a variable”.